### PR TITLE
feat: Segmented value type support generics

### PR DESCRIPTION
**Pull request by madocto**
[[中文版模板 / Chinese template](https://github.com/ant-design/ant-design/blob/master/.github/PULL_REQUEST_TEMPLATE/pr_cn.md?plain=1)]

### 🤔 This is a ...

- [x] New feature
- [ ] Bug fix
- [ ] Site / documentation update
- [ ] Demo update
- [ ] Component style update
- [ ] TypeScript definition update
- [ ] Bundle size optimization
- [ ] Performance optimization
- [ ] Enhancement feature
- [ ] Internationalization
- [ ] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Workflow
- [ ] Other (about what?)

### 🔗 Related issue link
close: #46234 

### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English |    Segmented `value` type support generics.    |
| 🇨🇳 Chinese |  Segmented `value` 类型支持泛型。    |

### ☑️ Self-Check before Merge

⚠️ Please check all items below before requesting a reviewing. ⚠️

- [x] Doc is updated/provided or not needed
- [x] Demo is updated/provided or not needed
- [x] TypeScript definition is updated/provided or not needed
- [x] Changelog is provided or not needed

### CODE REVIEW

File: components/segmented/index.tsx [hunk 1]
```diff
@@ -111,6 +117,14 @@ const Segmented = React.forwardRef<HTMLDivElement, SegmentedProps>((props, ref)
   );
 });
 
+const Segmented = InternalSegmented as (<ValueType>(
+  props: SegmentedProps<ValueType> & {
```
Comment (li-jia-nan) [inline components/segmented/index.tsx:121]:
```suggestion
  props: SegmentedProps<ValueType> & React.RefAttributes<HTMLDivElement>;
```

File: components/segmented/index.tsx [hunk 2]
```diff
@@ -111,6 +117,14 @@ const Segmented = React.forwardRef<HTMLDivElement, SegmentedProps>((props, ref)
...
+    ref?: React.ForwardedRef<HTMLDivElement>;
+  },
+) => ReturnType<typeof InternalSegmented>) & {
```
Comment (li-jia-nan) [inline components/segmented/index.tsx:124]:
```suggestion
) => ReturnType<typeof InternalSegmented>) & Pick<React.FC, "displayName">;
```

**Comment by afc163** [comment]
把 demo 代码也修改掉吧，展示如何使用这个泛型。

**Comment by crazyair** [comment]
https://github.com/react-component/segmented/pull/181 我这个改完了，还没合

File: components/segmented/demo/basic.tsx [hunk 3]
```diff
@@ -2,7 +2,12 @@ import React from 'react';
 import { Segmented } from 'antd';
 
 const Demo: React.FC = () => (
-  <Segmented options={['Daily', 'Weekly', 'Monthly', 'Quarterly', 'Yearly']} />
+  <Segmented<string | number | boolean>
+    options={['Daily', 'Weekly', 2021, 2022, true, false]}
+    onChange={(value) => {
+      console.log(value);
+    }}
+  />
```
Comment (afc163) [inline components/segmented/demo/basic.tsx:10]: 别改 demo，可以在 tests/type.test.tsx 里加个用例。

Comment (afc163) [inline components/segmented/demo/basic.tsx:10]:
```suggestion
  <Segmented<string>
    options={['Daily', 'Weekly', 'Monthly', 'Quarterly', 'Yearly']}
    onChange={(value) => {
      console.log(value); // string
    }}
  />
```

File: components/segmented/demo/basic.tsx [hunk 4]
```diff
@@ -2,7 +2,12 @@ import React from 'react';
 import { Segmented } from 'antd';
 
 const Demo: React.FC = () => (
-  <Segmented options={['Daily', 'Weekly', 'Monthly', 'Quarterly', 'Yearly']} />
+  <Segmented<string>
+    options={['Daily', 'Weekly', 'Monthly', 'Quarterly', 'Yearly']}
+    onChange={(value) => {
+      console.log(value); // string
+    }}
+  />
```
Comment (afc163) [inline components/segmented/demo/basic.tsx:10]: 我改回 string 的例子了，只需要加个 type 示例就行。

需要你帮忙回滚一下 snapshot。

**Review by afc163 — APPROVED** [review]

**Pull request merged**